module/zfs/dbuf.c

```diff
@@ -289,11 +356,22 @@ dbuf_hash_insert(dmu_buf_impl_t *db)
 		}
 	}
 
+	if (depth > 0) {
+		DBUF_STAT_BUMP(hash_chains);
+		if (depth == 1)
+			DBUF_STAT_BUMP(hash_collisions);
+		DBUF_STAT_MAX(hash_chain_max, depth);
+	}
+
 	mutex_enter(&db->db_mtx);
 	db->db_hash_next = h->hash_table[idx];
 	h->hash_table[idx] = db;
 	mutex_exit(DBUF_HASH_MUTEX(h, idx));
 	atomic_inc_64(&dbuf_hash_count);
+	DBUF_STAT_SET(hash_elements, dbuf_hash_count);
```

Comment:
Rather than always keep these two versions of this in sync I think it would make more sense to add a `ks_update` function.  Take a look at `arc_kstat_update` as an example.  This would allow you to remove entirely the `DBUF_STAT_SET` macro and simply dynamically generate these values when requested.